### DISCUSSION TITLE
Convert Images to .webp or avif on upload automaticaly on Joomla

### DISCUSSION THREAD

**Discussion by Jezjames** [joomla/joomla-cms#42822] [2. Cold Features]
Is your feature request related to a problem? Please describe.
For easy image optimization and SEO optimization.
Can the project edit the image uploader to alow the images uploaded to .webp or .avif file formats which are more compresed, load faster and better optizied.

Describe the solution you'd like
Edit or allow the joomla uploader to convert images to .webp or .avif on upload

**Comment by brianteeman** [comment]
I assume that you mean avif format not avi which is a video format,.

As far as I am aware there are several extensions that already do this that are available on the Extension Directory

**Reply by Jezjames** [reply]
Hi 
Sorry about that I meant .avif format which offers better compresion than the popular .webp extension

Thanks for pointing out the error.

**Reply by Jezjames** [reply]
Note just to add AVIF is the image version of the popular AV1 video format.

**Comment by niosme** [comment]
It would be nice the images for each component to be able to be converted , resize and compressed or define a specific aspect ratio to be uploaded on image upload. Most extensions do this after on a cron which isnt the best soliution.
I have a solution for my components and i can help on that and provide my code for all this. I would be happy to help with that :)

**Comment by dgrammatiko** [comment]
You could use some 3rd party plugin for this, ie: https://responsive-images.dgrammatiko.dev (not a paid ext)

**Reply by niosme** [reply]
Yes but it does it with crons or once the page loaded from someone. Its not while we upload the image to be converted, resized , renamed. Also i forgot the rename of the image file or remove spaces from images while uploading. A lot of people uploading pictures with spaces or with unrecognized characters.

The plugin is great but does it after which produces some loads on the user that triggers tha main taask of images creation.Not while the user uploads the image. Also there in not plugin event for a developer to handle the media file while its uploaded or before uploaded to change something like the name or whatever.

**Reply by dgrammatiko** [reply]
> Yes but it does it with crons or once the page loaded from someone

On upload

> The plugin is great but does it after which produces some loads on the user that triggers tha main taask of images creation.

It does that ONLY for non existing optimized images. There's also a CLI you can run to create the optimized version of all your existing images.

> Also there in not plugin event for a developer to handle the media file while its uploaded or before uploaded to change something like the name or whatever.

Where do you get that? the plugin is doing the work on upload: https://github.com/ttc-freebies/plugin-responsive-images/blob/29951a1ff7fe75a3a727ce3cc9993dfc7991dc43/src/plugins/media-action/responsive/language/en-GB/plg_media-action_responsive.ini#L2

OnAfterSave: https://github.com/ttc-freebies/plugin-responsive-images/blob/29951a1ff7fe75a3a727ce3cc9993dfc7991dc43/src/plugins/media-action/responsive/src/Extension/Responsive.php#L50-L76

**Reply by niosme** [reply]
Thats totally fine then as it is possible for someone to make it as you did.
It could be the best solution if there was an option in a future version to remove the original image, rename and keep the optimized only to prevent the space of very big images.

If we delete the main photo and the plugin manually for some reason the default functionallity will break, right?

**Comment by 4digitalworld** [comment]
Joomla doesn't have any options for image optimization, your request is similar to mine https://github.com/joomla/joomla-cms/issues/27813 with adding a field for Retina (2K, 4K) images.

It's 2024, everyone has a smartphone, TV and some laptops already have 2K or 4K resolution, but @Joomla doesn't have options for such screens!

**Reply by niosme** [reply]
https://magazine.joomla.org/all-issues/july-2020/j4-the-new-media-manager

it has some futures inside media manager

**Reply by dgrammatiko** [reply]
> It's 2024, everyone has a smartphone, TV and some laptops already have 2K or 4K resolution, but @joomla doesn't have options for such screens!


Sorry but you're wrong here. Joomla is a generic CMS, it doesn't have to necessarily optimize images to your preferences as your own preferences might be completely different than another user. That's the reason plugins exist. In the case of my own plugin I expect the users to upload the image to their best resolution and then the plugin (automatically) creates avif,webp and either jpeg or png (as the final fallback). Could I have done more, like 2x or whatever? Ofcourse I could but that's too specific, most people just want to serve an optimal (size/quality) image, the DPI part was requested only by photographers (also the storage needed is getting exponentially bigger the more formats/DPIs you're adding)

In sort what you're complaining about could already be done with a 3PD plugin but the core could not cover all the possible use cases of the CMS...

**Reply by 4digitalworld** [reply]
> 
> I expect the users to upload the image to their best resolution and then the plugin (automatically) creates avif,webp and either jpeg or png

all i want is to be able to specify alternative versions of the same image, for example for `2K`, `4K`, `8K` screens and for alternative formats like `webp`. 
i don't use the `Images and Links` tab because the functionality of that tab doesn't matter to me, it's too poor in functionality.

**Reply by dgrammatiko** [reply]
What you're asking for seems like a custom paid plugin. There are plenty devs that would be happy to help you with that

**Reply by 4digitalworld** [reply]
> What you're asking for seems like a custom paid plugin. There are plenty devs that would be happy to help you with that

I'm just thinking about what will happen to my website in 5 or 10 years, when most users will have `2K` or `4K` monitors, when most devices will have screen resolutions higher than today. 

Will I have to re-upload all the images on my website in 10 years to make them compatible with modern monitors and screens? 
This is a very big question and a very big problem that we all might think about today.

**Reply by dgrammatiko** [reply]
Probably you don't know this but SRGB colorspace images were the default for the web for ages. Only a year ago or so Chrome added support for P3 and other colorspaces. Why did I wrote this? Because it's the same exact situation, images from 10 years ago probably are lacking some of the modern stuff. BTW some formats are still SRGB only!

**Reply by dgrammatiko** [reply]
Also as I wrote before I ask users to upload the best version of their image and my code will create the efficient ones for the web. This way the uploaded images are always there, untouched in their best shape for any future changes. So I guess I'm doing it right(?)

**Reply by 4digitalworld** [reply]
> images from 10 years ago

**my website is more than 15 years old** and was built on `Joomla 1.0` and I updated my website every time a new version of Joomla was released. Now my website is using `Joomla 5.2` and has several thousand articles, several thousand images in articles and more than 200+ categories, more than 200+ tags and has different language versions of the site (via built-in multilingual feature in Joomla).

some images on my website **already look bad**, not clear, when I view my website on my smartphone, because my smartphone has a screen with a resolution of about `2K` and what will happen in 10 years when most screens will have a resolution of `4K` and more?
